README.md
# BHToast

![Swift](https://img.shields.io/badge/language-Swift-orange.svg)
[![CI Status](http://img.shields.io/travis/bbheck/BHToast.svg?style=flat)](https://travis-ci.org/bbheck/BHToast)
[![Version](https://img.shields.io/cocoapods/v/BHToast.svg?style=flat)](http://cocoapods.org/pods/BHToast)
[![License](https://img.shields.io/cocoapods/l/BHToast.svg?style=flat)](http://cocoapods.org/pods/BHToast)
[![Platform](https://img.shields.io/cocoapods/p/BHToast.svg?style=flat)](http://cocoapods.org/pods/BHToast)

## Requirements

- iOS 8.0+
- Xcode 7.0+

## Installation

BHToast is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "BHToast"
```

## Usage

<img src="https://cloud.githubusercontent.com/assets/2257294/12374113/f796500a-bc78-11e5-93bc-ac886c3d1f66.gif">

### Showing a `BHToast` (only message)

```swift
BHToast(view: view, message: "This is an example message.").show()
```
or
```swift
let toast = BHToast(view: view)

toast.message = "This is an example message."
toast.show()
```

### Showing a `BHToast` (with image)

```swift
BHToast(view: view, message: "This is an example message.", imageView: `yourImageView`).show()
```
or
```swift
let toast = BHToast(view: view, message: "This is an example message.")

toast.imageView = `yourImageView`
toast.show()
```

### Showing a BHToast (with options)

```swift
import BHToast

let options = BHToastOptions(
    duration: 10.0,
    borderWidth: 0.0,
    cornerRadius: 0.0
)

BHToast(view: view, message: "This is an example message.", options: options).show()
```

### Customizing

`BHToastViewTag` is used to guarantee only one instance in view.
If necessary, change it before first `BHToast` instance.

You can create a `BHToastOptions` instance with these properties:

* The time that `BHToast` stays in `UIView`. *Default: 5.0*
```swift
duration: NSTimeInterval
```

* The animation time (in seconds). *Default: 0.4*
```swift
animationDuration: NSTimeInterval
```

* Toast background color. *Default: UIColor.lightGrayColor()*
```swift
backgroundColor: UIColor
```

* Border color. *Default: UIColor.darkGrayColor()*
```swift
borderColor: UIColor
```

* Border width. *Default: 1.0*
```swift
borderWidth: CGFloat
```

* Corner radius. *Default: 5.0*
```swift
cornerRadius: CGFloat
```

* The margin. *Default: 8.0*
```swift
margin: CGFloat
```

* The content insets. *Default: UIEdgeInsetsMake(8.0, 8.0, 8.0, 8.0)*
```swift
contentInsets: UIEdgeInsets
```

* The min height. *Default: 30.0*
```swift
minHeight: CGFloat
```

* The max height. *Default: 50.0*
```swift
maxHeight: CGFloat
```

* The message alignment. *Default: NSTextAlignment.Center*
```swift
messageAlignment: NSTextAlignment
```

* The message color. *Default: UIColor.whiteColor()*
```swift
messageColor: UIColor
```

* The message font. *Default: UIFont.systemFontOfSize(14.0)*
```swift
messageFont: UIFont
```

* The view position. *Default: BHToastPosition.bottom*
```
position: BHToastPosition
```

* The image position (if set). *Default: BHToastImagePosition.Left*
```swift
imagePosition: BHToastImagePosition
```

**All those properties have a default value. Change them to apply the customizations to all new `BHToast` instances.**

## Author

Bruno Hecktheuer, bbheck@gmail.com

## License

BHToast is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log

## 0.4.0 (2017-07-09)

### Enhancements

- Add two new positions. Now you can show the `BHToast` on the top, middle or bottom.
- Create a `BHToastUtils` to access the top most view controller.
- Change properties name (`bottomOffset` -> `margin`, `defaultBottomOffset` -> `defaultMargin`).

## 0.3.1 (2016-02-28)

### Enhancements

- Add support to add/change the message and the imageView.
- Create a generic method to manage the `BHToast` constraints.
- Improve some documentation.

### Bug fixed

- Fix when the same `BHToast` instance called more than once the `show` method.


## 0.3.0 (2016-01-17)

### Enhancements

- Add image support in two positions (Left and Right).
- Create enum `BHToastImagePosition`.
- Add `messageAlignment` in `BHToastOptions`.
- Change `messagePadding` to `contentInsets` in `BHToastOptions`.
- Transfer `animationDuration` to `BHToastOptions`.


## 0.2.0 (2016-01-16)

### Enhancements

- Add the parameter options (`BHToastOptions`) in init method to a specific customization.
- Change the way how to customize the `BHToast`. In int method or change the `BHToastOptions` default values.


## 0.1.0 (2016-01-13)

### Enhancements

- Change `height` to `minHeight` and `maxHeight` properties in `BHToastOptions`
- Add constraints to `minHeight` and `maxHeight` in `BHToast`


## 0.1.0 (2016-01-12)

### Enhancements

- Created example project
- Support CocoaPods


Example/Tests/Tests.swift
// https://github.com/Quick/Quick

import Quick
import Nimble
import BHToast

class TableOfContentsSpec: QuickSpec {
    override func spec() {
        describe("these will fail") {

            it("can do maths") {
                expect(1) == 2
            }

            it("can read") {
                expect("number") == "string"
            }

            it("will eventually fail") {
                expect("time").toEventually( equal("done") )
            }
            
            context("these will pass") {

                it("can do maths") {
                    expect(23) == 23
                }

                it("can read") {
                    expect("🐮") == "🐮"
                }

                it("will eventually pass") {
                    var time = "passing"

                    dispatch_async(dispatch_get_main_queue()) {
                        time = "done"
                    }

                    waitUntil { done in
                        NSThread.sleepForTimeInterval(0.5)
                        expect(time) == "done"

                        done()
                    }
                }
            }
        }
    }
}


Example/BHToast_Tests/BHToast_Tests.swift
//
//  BHToast_Tests.swift
//  BHToast_Tests
//
//  Created by Bruno Hecktheuer on 9/23/16.
//  Copyright © 2016 CocoaPods. All rights reserved.
//

import XCTest

@testable import BHToast

class BHToast_Tests: XCTestCase {
    fileprivate var toast: BHToast!
    
    fileprivate let message = "This is an example message"
    fileprivate let margin: CGFloat = 40.0
    fileprivate let width: CGFloat = 300.0
    fileprivate let minHeight: CGFloat = 40.0
    fileprivate let maxHeight: CGFloat = 60.0
    fileprivate let contentInsets = UIEdgeInsets(top: 8.0, left: 8.0, bottom: 8.0, right: 8.0)
    
    fileprivate var window: UIWindow {
        return UIApplication.shared.windows.first!
    }
    
    fileprivate var view: UIView {
        return window.rootViewController!.view
    }
    
    override func setUp() {
        super.setUp()
        
        BHToastOptions.defaultAnimationDuration = 0.1
    }
    
    override func tearDown() {
        super.tearDown()
    }
    
    func testView_show() {
        toast = BHToast(view: view, message: message)
        toast.show()
        
        sleep()
        
        XCTAssertTrue(view.subviews.contains(toast))
    }
    
    func testView_hide() {
        toast = BHToast(view: view, message: message)
        toast.show()
        
        sleep()
        
        XCTAssertTrue(view.subviews.contains(toast))
    }
    
    func testView_width() {
        toast = BHToast(view: view, message: message)
        toast.show()
        
        sleep()
        
        toast.hide()
        
        sleep()
        
        XCTAssertFalse(view.subviews.contains(toast))
    }
    
    func testView_height() {
        toast = BHToast(
            view: view,
            message: message,
            options: BHToastOptions(
                minHeight: minHeight,
                maxHeight: maxHeight
            )
        )
        toast.show()
        
        sleep()
        
        XCTAssertGreaterThanOrEqual(toast.frame.height, minHeight)
        XCTAssertLessThanOrEqual(toast.frame.height, maxHeight)
    }
    
    func testView_topPosition() {
        toast = BHToast(
            view: view,
            message: message,
            options: BHToastOptions(
                margin: margin,
                position: .top
            )
        )
        toast.show()
        
        sleep()
        
        XCTAssertEqual(
            toast.frame.origin.y,
            margin
        )
    }
    
    func testView_middlePosition() {
        toast = BHToast(
            view: view,
            message: message,
            options: BHToastOptions(
                margin: margin,
                position: .middle
            )
        )
        toast.show()
        
        sleep()
        
        XCTAssertEqual(
            toast.center,
            view.center
        )
    }
    
    func testView_bottomPosition() {
        toast = BHToast(
            view: view,
            message: message,
            options: BHToastOptions(
                margin: margin,
                position: .bottom
            )
        )
        toast.show()
        
        sleep()
        
        XCTAssertEqual(
            view.frame.height - (toast.frame.origin.y + toast.frame.height),
            margin
        )
    }
    
    func testMessageLabel_text() {
        toast = BHToast(view: view, message: message)
        toast.show()
        
        sleep()
        
        XCTAssertEqual(toast.messageLabel.text, message)
    }
    
    func testMessageLabel_edges() {
        toast = BHToast(
            view: view,
            message: message,
            options: BHToastOptions(
                contentInsets: contentInsets
            )
        )
        toast.show()
        
        sleep()
        
        XCTAssertEqual(
            toast.messageLabel.frame.origin.y,
            contentInsets.top
        )
        
        XCTAssertEqual(
            toast.frame.height -
                (toast.messageLabel.frame.origin.y +
                    toast.messageLabel.frame.height),
            contentInsets.bottom
        )
        
        XCTAssertEqual(
            toast.messageLabel.frame.origin.y,
            contentInsets.left
        )
        
        XCTAssertEqual(
            toast.frame.width -
                (toast.messageLabel.frame.origin.x +
                    toast.messageLabel.frame.width),
            contentInsets.right
        )
    }
    
    func testImageView_height() {
        toast = BHToast(
            view: view,
            message: message,
            imageView: UIImageView(
                image: UIImage(
                    named: "SuccessIcon"
                )
            ),
            options: BHToastOptions(
                minHeight: minHeight,
                maxHeight: maxHeight,
                imagePosition: .left
            )
        )
        toast.show()
        
        sleep()
        
        XCTAssertEqual(
            (toast.imageView!.frame.origin.y * 2) + toast.imageView!.frame.height,
            minHeight
        )
    }
    
    fileprivate func sleep(duration: TimeInterval = 0.2) {
        RunLoop.current.run(
            until: Date().addingTimeInterval(duration)
        )
    }
    
}


Pod/Classes/BHToast.swift
//
//  BHToast.swift
//  Pods
//
//  Created by Bruno Hecktheuer on 1/11/16.
//
//

import UIKit

/// The BHToast class defines a custom UIView with a message.
@available(iOS 8.0, *)
open class BHToast: UIView {
    
    // MARK: - Properties
    
    /// The parent UIView that shows the BHToast.
    private let view: UIView
    
    /// The display message.
    open var message: String
    
    /// The display imageView
    open var imageView: UIImageView?
    
    /// The BHToastOptions
    private let options: BHToastOptions
    
    /// The BHToast width (fixed 300.0).
    private let width: CGFloat = 300.0
    
    /// The UILabel message
    internal let messageLabel = UILabel()
    
    /// The timer that hide the BHToast.
    private var timer = Timer()
    
    // MARK: - Init methods
    
    /**
     Custom init method.

     Create an instance of BHToast to attach in an UIView.
     You should set a message and can add an image view to display.

     - parameter view:              The UIView that shows the BHToast.
     - parameter message:           The display message.
     - parameter imageView:         The display image view.
     - parameter options:           The BHToastOptions instance.
    */
    public init(view: UIView? = nil,
                message: String = "",
                imageView: UIImageView? = nil,
                options: BHToastOptions = BHToastOptions()) {
        
        self.view = view ?? BHToastUtils.topViewController.view
        self.message = message
        self.imageView = imageView
        self.options = options
        
        super.init(frame: CGRect.zero)
        
        setupViewProperties()
    }

    @available(*, unavailable, message: "required init is unavailable")
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: - Setup methods
    
    /**
     Configures view constraints and calls the other setup methods.
    */
    private func setupProperties() {
        addWidthConstraintToElement(self, rule: "\(width)")
        addHeightConstraintToElement(self, rule: ">=\(options.minHeight)")
        addHeightConstraintToElement(self, rule: "<=\(options.maxHeight)")
        
        addConstraintFrom( // Align Center X
            view, fromAttribute: .centerX,
            to: self, toAttribute: .centerX
        )
        
        setupViewPosition()
        
        setupLayerProperties()
        
        if let _ = imageView {
            setupImageView()
        }
        
        setupMessageLabel()
    }
    
    private func setupViewPosition() {
        switch options.position {
        case .bottom:
            addConstraintFrom( // Bottom Margin
                view, fromAttribute: .bottom,
                to: self, toAttribute: .bottom,
                value: options.margin
            )
        case .middle:
            addConstraintFrom( // Align Center Y
                view, fromAttribute: .centerY,
                to: self, toAttribute: .centerY
            )
        case .top:
            addConstraintFrom( // Top Margin
                self, fromAttribute: .top,
                to: view, toAttribute: .top,
                value: options.margin
            )
        }
    }
    
    /**
     Sets the view properties.
    */
    private func setupViewProperties() {
        tag = BHToastViewTag
        
        alpha = 0.0
        backgroundColor = options.backgroundColor
        
        translatesAutoresizingMaskIntoConstraints = false
    }
    
    /**
     Sets the layer properties.
    */
    private func setupLayerProperties() {
        layer.borderWidth = options.borderWidth
        layer.borderColor = options.borderColor.cgColor
        
        layer.cornerRadius = options.cornerRadius
        
        layer.masksToBounds = true
    }
    
    /**
     Sets the UIImageView properties.
    */
    private func setupImageView() {
        addSubview(imageView!)
        
        imageView!.translatesAutoresizingMaskIntoConstraints = false
        
        addConstraintFrom( // Align Center Y
            self, fromAttribute: .centerY,
            to: imageView!, toAttribute: .centerY
        )
        
        if options.imagePosition == .left {
            addConstraintFrom( // Left Margin
                imageView!, fromAttribute: .left,
                to: self, toAttribute: .left,
                value: options.contentInsets.left
            )
        } else {
            addConstraintFrom( // Right Margin
                self, fromAttribute: .right,
                to: imageView!, toAttribute: .right,
                value: options.contentInsets.right
            )
        }
        
        addHeightConstraintToElement(
            imageView!,
            rule: "\(options.minHeight - ((options.contentInsets.top + options.contentInsets.bottom) / 2))"
        )
        
        addConstraintFrom( // Ratio Constraint
            imageView!, fromAttribute: .height,
            to: imageView!, toAttribute: .width,
            multiplier: imageView!.frame.height / imageView!.frame.width
        )
    }
    
    /**
     Sets the message label properties.
     */
    private func setupMessageLabel() {
        messageLabel.font = options.messageFont
        
        messageLabel.numberOfLines = 0
        
        messageLabel.text = message
        messageLabel.textAlignment = options.messageAlignment
        messageLabel.textColor = options.messageColor
        
        addSubview(messageLabel)
        
        messageLabel.translatesAutoresizingMaskIntoConstraints = false
        
        addConstraintFrom( // Top Margin
            messageLabel, fromAttribute: .top,
            to: self, toAttribute: .top,
            value: options.contentInsets.top
        )
        
        addConstraintFrom( // Bottom Margin
            self, fromAttribute: .bottom,
            to: messageLabel, toAttribute: .bottom,
            value: options.contentInsets.bottom
        )
        
        if imageView == nil {
            addConstraintFrom( // Left Margin
                messageLabel, fromAttribute: .left,
                to: self, toAttribute: .left,
                value: options.contentInsets.left
            )
            
            addConstraintFrom( // Right Margin
                self, fromAttribute: .right,
                to: messageLabel, toAttribute: .right,
                value: options.contentInsets.right
            )
        } else {
            addConstraintFrom( // Left Margin
                messageLabel,
                fromAttribute: .left,
                to: options.imagePosition == .left ? imageView! : self,
                toAttribute: options.imagePosition == .left ? .right : .left,
                value: options.contentInsets.left
            )
            
            addConstraintFrom( // Right Margin
                options.imagePosition == .right ? imageView! : self,
                fromAttribute: options.imagePosition == .right ? .left : .right,
                to: messageLabel,
                toAttribute: .right,
                value: options.contentInsets.right
            )
        }
        
    }
    
    // MARK: - Scheduled method
    
    /**
     Starts the timer that hide the BHToast.
     */
    private func scheduledHideEvent() {
        timer = Timer.scheduledTimer(
            timeInterval: options.duration, target: self,
            selector: #selector(hide), userInfo: nil,
            repeats: false
        )
    }
    
    // MARK: - Private show method
    
    /**
     Shows the view with animation.
     */
    private func showWithAnimation() {
        view.addSubview(self)
        
        setupProperties()
        
        UIView.animate(
            withDuration: options.animationDuration,
            animations: { () -> Void in
                self.alpha = 1.0
            }, completion: { (finish) -> Void in
                self.scheduledHideEvent()
            }
        )
    }
    
    // MARK: - Private hide method
    
    /**
     Hides the view with animation.

     - parameter showAgain: The flag that checks if is necessary to show again the view.
     */
    private func hideWithAnimation(completionHandler: @escaping () -> Void = {}) {
        UIView.animate(
            withDuration: options.animationDuration,
            animations: { () -> Void in
                self.alpha = 0.0
            }, completion: { (finish) -> Void in
                self.removeFromSuperview()
                self.timer.invalidate()
                completionHandler()
            }
        )
    }
    
    // MARK: - Public methods
    
    /**
     Hides the BHToast with fade animation.
     */
    open func hide() {
        hideWithAnimation()
    }
    
    /**
     Shows the BHToast with animation.
     
     * If BHToast has already been added to parent view, dismiss the BHToast to show again.
     * If some BHToast with BHToastViewTag already exists in parent view, dismiss it to show the new BHToast.
     * Otherwise, just show the BHToast with fade animation.
     */
    open func show() {
        if isDescendant(of: view) {
            hideWithAnimation(completionHandler: { () -> Void in
                self.showWithAnimation()
            })
        } else if let _view = view.viewWithTag(BHToastViewTag) as? BHToast {
            _view.hideWithAnimation(completionHandler: { () -> Void in
                self.showWithAnimation()
            })
        } else {
            showWithAnimation()
        }
    }
    
    // MARK: - Constraint methods
    
    /**
     Generic method to add a constraint.

     - parameter: from:             The "from" AnyObject.
     - parameter: fromAttribute:    The "from" NSLayoutAttribute.
     - parameter: relatedBy:        The NSLayoutRelation (default: .Equal)
     - parameter: to:               The "to" AnyObject.
     - parameter: toAttribute:      The "to" NSLayoutAttribute.
     - parameter: multiplier:       The multiplier to change the value (default: 1.0)
     - parameter: value:            The value set in the constraint (default: 0.0).
     - parameter: priority:         The constraint priority (default: 1000).
     */
    private func addConstraintFrom(_ from: AnyObject,
                                   fromAttribute: NSLayoutAttribute,
                                   relatedBy: NSLayoutRelation = .equal,
                                   to: AnyObject,
                                   toAttribute: NSLayoutAttribute,
                                   multiplier: CGFloat = 1.0,
                                   value: CGFloat = 0.0,
                                   priority: UILayoutPriority = 1000) {
        
        let constraint = NSLayoutConstraint(
            item: from,
            attribute: fromAttribute,
            relatedBy: relatedBy,
            toItem: to,
            attribute: toAttribute,
            multiplier: multiplier,
            constant: value
        )
        
        constraint.priority = priority
        view.addConstraint(constraint)
    }
    
    /**
     Add height constraint.

     - parameter element:   AnyObject
     - parameter rule:      String (examples: "<=200", ">300", "200")
     */
    private func addHeightConstraintToElement(_ element: AnyObject, rule: String) {
        element.addConstraints(
            NSLayoutConstraint.constraints(
                withVisualFormat: "V:[element(\(rule))]",
                options: NSLayoutFormatOptions(rawValue: 0),
                metrics: nil,
                views: [
                    "element": element
                ]
            )
        )
    }
    
    /**
     Add width constraint

     - parameter element:   AnyObject
     - parameter rule:      String (examples: "<=200", ">300", "200")
     */
    private func addWidthConstraintToElement(_ element: AnyObject, rule: String) {
        element.addConstraints(
            NSLayoutConstraint.constraints(
                withVisualFormat: "H:[element(\(rule))]",
                options: NSLayoutFormatOptions(rawValue: 0),
                metrics: nil,
                views: [
                    "element": element
                ]
            )
        )
    }
    
}


Pod/Classes/BHToastOptions.swift
//
//  BHToastOptions.swift
//  Pods
//
//  Created by Bruno Hecktheuer on 1/12/16.
//
//

import UIKit

/**
 The BHToastViewTag is used to guarantee only one instance.

 *If necessary, change it before create the first BHToast instance*.
*/
public var BHToastViewTag = 3091990

/// Represents the view position (Bottom, Middle, Top)
public enum BHToastPosition {
    case bottom, middle, top
}

/// Represents the image position (Left or Right)
public enum BHToastImagePosition {
    case left, right
}

/// The struct of customization options.
public struct BHToastOptions {
    
    // Default properties (change if necessary)
    public static var defaultDuration: TimeInterval = 5.0
    public static var defaultAnimationDuration: TimeInterval = 0.4
    public static var defaultBackgroundColor: UIColor = .lightGray
    public static var defaultBorderColor: UIColor = .darkGray
    public static var defaultBorderWidth: CGFloat = 1.0
    public static var defaultCornerRadius: CGFloat = 5.0
    /// Only applies when the position is top or bottom.
    public static var defaultMargin: CGFloat = 8.0
    public static var defaultContentInsets = UIEdgeInsets(
        top: 8.0, left: 8.0, bottom: 8.0, right: 8.0
    )
    public static var defaultMinHeight: CGFloat = 30.0
    public static var defaultMaxHeight: CGFloat = 50.0
    public static var defaultMessageAlignment: NSTextAlignment = .center
    public static var defaultMessageColor: UIColor = .white
    public static var defaultMessageFont = UIFont.systemFont(ofSize: 14.0)
    public static var defaultPosition: BHToastPosition = .bottom
    public static var defaultImagePosition: BHToastImagePosition = .left
    
    // MARK: - Properties
    
    /// The duration that the Toast stays in View (*in seconds*).
    public let duration: TimeInterval
    
    /// The animation time (*in seconds*).
    public let animationDuration: TimeInterval
    
    /// Toast background color.
    public let backgroundColor: UIColor
    
    // Border options
    public let borderColor: UIColor
    public let cornerRadius: CGFloat
    public let borderWidth: CGFloat
    
    // View size options
    public let margin: CGFloat
    public let contentInsets: UIEdgeInsets
    public let minHeight: CGFloat
    public let maxHeight: CGFloat
    
    // Message options
    public let messageAlignment: NSTextAlignment
    public let messageColor: UIColor
    public let messageFont: UIFont
    
    // View position
    public let position: BHToastPosition
    
    // Image options
    public let imagePosition: BHToastImagePosition
    
    // MARK: - Init method
    
    /**
     Create an instance of BHToastOptions with some customizations.
    
     If necessary, change the default values to apply the customization for all new BHToastOptions instance.
     */
    public init(duration: TimeInterval = defaultDuration,
                animationDuration: TimeInterval = defaultAnimationDuration,
                backgroundColor: UIColor = defaultBackgroundColor,
                borderColor: UIColor = defaultBorderColor,
                borderWidth: CGFloat = defaultBorderWidth,
                cornerRadius: CGFloat = defaultCornerRadius,
                margin: CGFloat = defaultMargin,
                contentInsets: UIEdgeInsets = defaultContentInsets,
                minHeight: CGFloat = defaultMinHeight,
                maxHeight: CGFloat = defaultMaxHeight,
                messageAlignment: NSTextAlignment = defaultMessageAlignment,
                messageColor: UIColor = defaultMessageColor,
                messageFont: UIFont = defaultMessageFont,
                position: BHToastPosition = defaultPosition,
                imagePosition: BHToastImagePosition = defaultImagePosition) {
        
        self.duration = duration
        self.animationDuration = animationDuration
        self.backgroundColor = backgroundColor
        self.borderColor = borderColor
        self.borderWidth = borderWidth
        self.cornerRadius = cornerRadius
        self.margin = margin
        self.contentInsets = contentInsets
        self.minHeight = minHeight
        self.maxHeight = maxHeight
        self.messageAlignment = messageAlignment
        self.messageColor = messageColor
        self.messageFont = messageFont
        self.position = position
        self.imagePosition = imagePosition
    }
    
}


Example/BHToast/ViewController.swift
//
//  ViewController.swift
//  BHToast
//
//  Created by Bruno Hecktheuer on 01/11/2016.
//  Copyright (c) 2016 Bruno Hecktheuer. All rights reserved.
//

import UIKit

import BHToast

class ViewController: UIViewController {
    
    lazy var toast: BHToast = BHToast()
    
    let messages = [
        "Lorem ipsum dolor sit amet, vim liber iudicabit assentior ut, mel ad aliquid omittantur.",
        "No referrentur adversarium nec.",
        "Usu ne stet autem scripta, has partem forensibus eu.",
        "Ad eos iusto mnesarchum, latine rationibus deterruisset id qui.",
        "Et eum purto malorum alienum, vim ne animal salutatus, cu blandit scaevola officiis vis.",
        "Usu augue oporteat cu, vide vitae an mel.",
        "Ad vix clita mediocritatem, nam in debet doming labitur, dicant molestie necessitatibus mei ne.",
        "Sea nisl atqui eu, sed eu iusto vitae, has ea modo voluptua dissentiet."
    ]
    
    var randomMessage: String {
        let index = Int(arc4random_uniform(UInt32(messages.count - 1)))
        return messages[index]
    }

}

// MARK: - Actions

extension ViewController {
    
    @IBAction func showToastWithImage(_ sender: UIButton) {
        let isSuccess = arc4random_uniform(10) % 2 == 0
        
        BHToast(
            view: view,
            message: randomMessage,
            imageView: UIImageView(
                image: UIImage(
                    named: isSuccess ? "SuccessIcon" : "ErrorIcon"
                )
            ),
            options: BHToastOptions(
                margin: 44.0,
                minHeight: 40.0,
                position: .top,
                imagePosition: isSuccess ? .left : .right
            )
        ).show()
    }
    
    @IBAction func showToastAction(_ sender: UIButton) {
        toast.message = randomMessage
        toast.show()
    }
    
}



Example/BHToast_ExampleTests/BHToast_ExampleTests.swift
//
//  BHToast_ExampleTests.swift
//  BHToast_ExampleTests
//
//  Created by Bruno Hecktheuer on 3/24/16.
//  Copyright © 2016 CocoaPods. All rights reserved.
//

import XCTest

@testable import BHToast

class BHToast_ExampleTests: XCTestCase {
    private var toast: BHToast!
    
    private let message = "This is an example message"
    private let bottomOffset: CGFloat = 40.0
    private let width: CGFloat = 300.0
    private let minHeight: CGFloat = 40.0
    private let maxHeight: CGFloat = 60.0
    private let contentInsets = UIEdgeInsets(top: 8.0, left: 8.0, bottom: 8.0, right: 8.0)
    
    private var window: UIWindow {
        return UIApplication.sharedApplication().windows.first!
    }
    
    private var view: UIView {
        return window.rootViewController!.view
    }
    
    override func setUp() {
        super.setUp()
        
        BHToastOptions.defaultAnimationDuration = 0.1
    }
    
    override func tearDown() {
        super.tearDown()
    }
    
    func testView_show() {
        toast = BHToast(view: view, message: message)
        toast.show()
        
        sleep()
        
        XCTAssertTrue(view.subviews.contains(toast))
    }
    
    func testView_hide() {
        toast = BHToast(view: view, message: message)
        toast.show()
        
        sleep()
        
        XCTAssertTrue(view.subviews.contains(toast))
    }
    
    func testView_width() {
        toast = BHToast(view: view, message: message)
        toast.show()
        
        sleep()
        
        toast.hide()
        
        sleep()
        
        XCTAssertFalse(view.subviews.contains(toast))
    }
    
    func testView_height() {
        toast = BHToast(
            view: view,
            message: message,
            options: BHToastOptions(
                minHeight: minHeight,
                maxHeight: maxHeight
            )
        )
        toast.show()
        
        sleep()
        
        XCTAssertGreaterThanOrEqual(toast.frame.height, minHeight)
        XCTAssertLessThanOrEqual(toast.frame.height, maxHeight)
    }
    
    func testView_bottomOffset() {
        toast = BHToast(
            view: view,
            message: message,
            options: BHToastOptions(
                bottomOffset: bottomOffset
            )
        )
        toast.show()

        sleep()
        
        XCTAssertGreaterThanOrEqual(
            view.frame.height -
            (toast.frame.origin.y +
            toast.frame.height),
            bottomOffset
        )
    }
    
    func testMessageLabel_text() {
        toast = BHToast(view: view, message: message)
        toast.show()
        
        sleep()
        
        XCTAssertEqual(toast.messageLabel.text, message)
    }
    
    func testMessageLabel_edges() {
        toast = BHToast(
            view: view,
            message: message,
            options: BHToastOptions(
                contentInsets: contentInsets
            )
        )
        toast.show()
        
        sleep()
        
        XCTAssertEqual(
            toast.messageLabel.frame.origin.y,
            contentInsets.top
        )
        
        XCTAssertEqual(
            toast.frame.height -
            (toast.messageLabel.frame.origin.y +
            toast.messageLabel.frame.height),
            contentInsets.bottom
        )
        
        XCTAssertEqual(
            toast.messageLabel.frame.origin.y,
            contentInsets.left
        )
        
        XCTAssertEqual(
            toast.frame.width -
            (toast.messageLabel.frame.origin.x +
            toast.messageLabel.frame.width),
            contentInsets.right
        )
    }
    
    func testImageView_height() {
        toast = BHToast(
            view: view,
            message: message,
            imageView: UIImageView(
                image: UIImage(
                    named: "SuccessIcon"
                )
            ),
            options: BHToastOptions(
                minHeight: minHeight,
                maxHeight: maxHeight,
                imagePosition: .Left
            )
        )
        toast.show()
        
        sleep()
        
        XCTAssertEqual(
            (toast.imageView!.frame.origin.y * 2) + toast.imageView!.frame.height,
            minHeight
        )
    }
    
    private func sleep(duration duration: NSTimeInterval = 0.2) {
        NSRunLoop.currentRunLoop().runUntilDate(
            NSDate().dateByAddingTimeInterval(duration)
        )
    }
    
}
